src/main/scala/is/hail/annotations/CodeOrdering.scala

```diff
@@ -1,187 +1,404 @@
-                              cmp := compare(ti.pointType, mb, r1, ti.loadEnd(r1, i1), r2, ti.loadEnd(r2, i2)),
-                              cmp.cne(0).mux(cmp,
-                                Code(
-                                  pinclude1 := ti.includeEnd(r1, i1),
-                                  pinclude1.cne(ti.includeEnd(r2, i2))
-                                    .mux(pinclude1.mux(1, -1), 0)))),
-                            0))))))),
-              0)))
-      case ts: TBaseStruct =>
-        var i = ts.size - 1
-        val leftDefined = mb.newLocal[Boolean]
-        val rightDefined = mb.newLocal[Boolean]
-        val cmp = mb.newLocal[Int]
-
-        var compNext = Code._empty[Unit]
-        while(i >= 0) {
-          compNext = Code(
-              leftDefined := ts.isFieldDefined(r1, o1, i),
-              rightDefined := ts.isFieldDefined(r2, o2, i),
-              (leftDefined && rightDefined).mux(
-                cmp := compare(ts.types(i), mb, r1, ts.loadField(r1, o1, i), r2, ts.loadField(r2, o2, i)),
-                leftDefined.ceq(rightDefined).mux(
-                  Code._empty,
-                  cmp := (if (missingGreatest) leftDefined.mux(-1, 1) else leftDefined.mux(1, -1)))),
-              cmp.ceq(0).mux(compNext, Code._empty))
-          i -= 1
-        }
-        Code(cmp := 0, compNext, cmp)
-
-      case tc: TContainer =>
-        val etype = tc.elementType
-
-        val a1 = mb.newLocal[Long]("it_ord_a1")
-        val a2 = mb.newLocal[Long]("it_ord_a2")
-        val l1 = mb.newLocal[Int]("it_ord_l1")
-        val l2 = mb.newLocal[Int]("it_ord_l2")
-        val lim = mb.newLocal[Int]("it_ord_lim")
-        val i = mb.newLocal[Int]("it_ord_i")
-        val cmp = mb.newLocal[Int]("it_ord_cmp")
-        val leftDefined = mb.newLocal[Boolean]
-        val rightDefined = mb.newLocal[Boolean]
-        asm4s.coerce[Int](Code(
-          a1 := o1,
-          a2 := o2,
-          l1 := tc.loadLength(r1, o1),
-          l2 := tc.loadLength(r2, o2),
-          lim := (l1 < l2).mux(l1, l2),
-          i := 0,
-          cmp := 0,
-          Code.whileLoop(cmp.ceq(0) && i < lim,
-            leftDefined := tc.isElementDefined(r1, a1, i),
-            rightDefined := tc.isElementDefined(r2, a2, i),
-            (leftDefined && rightDefined).mux(
-              cmp := compare(etype, mb,
-                r1, tc.loadElement(r1, a1, i),
-                r2, tc.loadElement(r2, a2, i)),
-              leftDefined.cne(rightDefined).mux(
-                cmp := (if (missingGreatest) leftDefined.mux(-1, 1) else leftDefined.mux(1, -1)),
-                Code._empty)),
-            i += 1),
-          cmp.ceq(0).mux(intCompare(l1, l2), cmp)))
-      case tc: ComplexType =>
-        compare(tc.representation, mb, r1, o1, r2, o2)
-      case _ =>
-        throw new UnsupportedOperationException(s"can't stage type: $typ")
+  def iterableOrdering(t: TArray, mb: EmitMethodBuilder): CodeOrdering = new CodeOrdering {
```

Comment:
can this be a TCollection?

Reply:
`TContainer`? It could. I kind of wanted to keep the the orderings for sets/dicts/arrays conceptually separate so we're assuming the physical layout of sets and dicts as sorted arrays in fewer places, but I can combine them if you want.

Reply:
that seems OK.